server/utils/verifiers/VerifierBase.ts

```diff
@@ -0,0 +1,350 @@
+import * as COL from '../../collections/interface';
+import * as LOG from '../logger';
+import * as DBAPI from '../../db';
+import * as CACHE from '../../cache';
+import * as COMMON from '@dpo-packrat/common';
+// import * as H from '../helpers';
+
+export type IdentifierDetails = {
+    identifier: DBAPI.Identifier | null,
+    identifierType: DBAPI.Vocabulary | null,
+    identifierTypeEnum: COMMON.eVocabularyID | null,
+};
+
+export type IdentifierList = {
+    preferred: IdentifierDetails | null,
+    edan: IdentifierDetails | null,
+    ark: IdentifierDetails | null,
+    details: IdentifierDetails[],  // complete list of all identifiers
+};
+
+export class VerifierBase {
+
+    constructor() {}
+
+    protected async getIdentifierType(identifier: DBAPI.Identifier): Promise<IdentifierDetails | null> {
+
+        // grab the identifier type object from the Packrat DB
+        const identifierType: DBAPI.Vocabulary | null = await DBAPI.Vocabulary.fetch(identifier.idVIdentifierType);
+        if(!identifierType){
+            LOG.error(`could not find identifier type in DB (identifier: ${identifier.idVIdentifierType} )`, LOG.LS.eTEST);
+            return null;
+        }
+
+        // pull from enumeration from the CACHE (vocabulary id -> enum)
+        const identifierTypeEnum: COMMON.eVocabularyID | undefined = await CACHE.VocabularyCache.vocabularyIdToEnum(identifier.idVIdentifierType);
+        if(identifierTypeEnum===undefined){
+            LOG.error(`could not find enumerator for identifier type (${identifier.idVIdentifierType}) in Cache`, LOG.LS.eTEST);
+            return null;
+        }
+
+        return { identifier, identifierType, identifierTypeEnum };
+    }
+
+    protected async getSubjectIdentifiers(subject: DBAPI.Subject, systemObject: DBAPI.SystemObject): Promise<IdentifierList | null> {
+
+        // structure to hold our results
+        const result: IdentifierList = { preferred: null, edan: null, ark: null, details: [] };
+
+        // grab the preferred identifier, if nothing then leave null so calling can decide action
+        if(subject.idIdentifierPreferred) {
+            const preferredIdentifier: DBAPI.Identifier | null = await DBAPI.Identifier.fetch(subject.idIdentifierPreferred);
+            if(!preferredIdentifier){
+                LOG.error(`subject's preferredId not found in the DB (id: ${subject.idSubject} | subject: ${subject.Name})`, LOG.LS.eTEST);
+                // subjectStats[i].isValid = false;
+            } else {
+                // grab our identifier details (type) and store it
+                const preferredIdentifierDetails: IdentifierDetails | null = await this.getIdentifierType(preferredIdentifier);
+                if(preferredIdentifierDetails) {
+                    result.preferred = preferredIdentifierDetails;
+                    result.details?.push(preferredIdentifierDetails);

```

Comment:
Why are you using the "?" in `result.details?.push`?  It looks like result.details is always defined above.